### PR TITLE
LBAC-34: Added UserRole Based Location Access

### PR DESCRIPTION
**Pull request by VANKINEENITAWRUN**
## Description of what I changed

Added a function`getUserAccessibleLocationUuidsWithUserPrivileges`  in utils to get accessible Locations for the user. will replace the current function `getUserAccessibleLocationUuid` once LBAC-33 gets merged. Making changes in `getUserAccessibleLocationUuid` we need to make changes in All classes using it and then will make conflicts once LBAC-33.

## Issue I worked on

see https://issues.openmrs.org/browse/LBAC-34

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java [hunk 1]
```diff
@@ -49,13 +54,23 @@ public void willStart() {
...
 		LocationUtils.createPersonAttributeTypeForLocation();
+		generateOpenMRSPrivilegeForLocations();
 	    createGlobalPropertyForLocationAttribute(LocationBasedAccessConstants.LOCATION_ATTRIBUTE_GLOBAL_PROPERTY_NAME,
 			    LocationBasedAccessConstants.PERSONATTRIBUTETYPE_UUID);
 	    createGlobalPropertyForLocationAttribute(LocationBasedAccessConstants.REF_APP_LOCATION_USER_PROPERTY_NAME,
 			    LocationBasedAccessConstants.LOCATION_USER_PROPERTY_NAME);
 	    log.info("Location Based Access Control Module started");
     }
 
+	private void generateOpenMRSPrivilegeForLocations() {
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java]: why we need OpenMRS in this method name? Let's go for a generic name

File: api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java [hunk 2]
```diff
@@ -49,13 +54,23 @@ public void willStart() {
...
+		UserService userService = Context.getUserService();
+		List<Location> openMrsLocations = Context.getLocationService().getAllLocations();
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java]: Simply `allActiveLocations`?

File: api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java [hunk 3]
```diff
@@ -49,13 +54,23 @@ public void willStart() {
...
+		for(Location location : openMrsLocations){
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java]: format this line, 
further is there any possibilities for `openMrsLocations` null?

File: api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java [hunk 4]
```diff
@@ -80,6 +85,31 @@ public static String getUserAccessibleLocationUuid(User authenticatedUser) {
...
+     */
+    public static List <String> getUserAccessibleLocationUuidsWithUserPrivileges(User authenticatedUser){
+        if (authenticatedUser == null) {
+            return null;
+        }
+        List <String> accessibleLocationUuids = new ArrayList<String>();
+        String accessibleLocationUuid = authenticatedUser.getUserProperty(LocationBasedAccessConstants.LOCATION_USER_PROPERTY_NAME);
+        if (StringUtils.isBlank(accessibleLocationUuid)) {
+            List<Location> openMrsLocations = Context.getLocationService().getAllLocations();
+            for(Location location : openMrsLocations){
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java]: format this line

File: api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java [hunk 5]
```diff
@@ -80,6 +85,31 @@ public static String getUserAccessibleLocationUuid(User authenticatedUser) {
...
+                if(authenticatedUser.hasPrivilege("LocationAccess "+location.getName())){
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java]: format this line too

File: api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java [hunk 6]
```diff
@@ -80,6 +85,31 @@ public static String getUserAccessibleLocationUuid(User authenticatedUser) {
...
+                    accessibleLocationUuids.add(location.getUuid());
+                }
+            }
+        }else{
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java]: format this line too

File: api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java (hunk 2 above)
Comment (VANKINEENITAWRUN) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java]: Currently, in user and patient Fragment controller, we are getting all the Locations, so here we are using all the Locations.
else we can change it to `getAllLocations(false)` to get only the non-retired one.

File: api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java (hunk 3 above)
Comment (VANKINEENITAWRUN) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java]: I don't think there will a point where OpenMRS with no Location

**Comment by dkayiwa** [comment]
@suthagar23 any more comments for this?

File: api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java (hunk 1 above)
Comment (VANKINEENITAWRUN) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java]: I think `generatePrivilegesForLocations` would be good.

Comment (suthagar23) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java]: Have you updated this PR with this name?

**Comment by suthagar23** [comment]
@dkayiwa Still some comments need to be addressed!

@VANKINEENITAWRUN Can you check the comments and follow JIRA guidelines also.

File: api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java (hunk 1 above)
Comment (VANKINEENITAWRUN) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java]: Yes Updated.

File: api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java [hunk 7]
```diff
@@ -80,6 +85,31 @@ public static String getUserAccessibleLocationUuid(User authenticatedUser) {
...
+     * @return accessible Location uuid
+     */
+    public static List <String> getUserAccessibleLocationUuidsWithUserPrivileges(User authenticatedUser){
+        if (authenticatedUser == null) {
+            return null;
+        }
+        List <String> accessibleLocationUuids = new ArrayList<String>();
+        String accessibleLocationUuid = authenticatedUser.getUserProperty(LocationBasedAccessConstants.LOCATION_USER_PROPERTY_NAME);
+        if (StringUtils.isBlank(accessibleLocationUuid)) {
+            List<Location> openMrsLocations = Context.getLocationService().getAllLocations();
```
Comment (VANKINEENITAWRUN) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java]: @suthagar23 the function `getUserAccessibleLocationUuidsWithUserPrivileges` is called from the `LocationServiceInterceptorAdvice` when the function is matched. here we are checking if privilege matches with any of the Location by `getAllLocations`  so it is being called again so this is intercepted by aop interceptor it is being recursive thereby giving a StackOverflow error.

**Comment by dkayiwa** [comment]
@suthagar23 what do you think of this?

File: api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java (hunk 7 above)
Comment (suthagar23) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java]: What could be the solution then?

Comment (VANKINEENITAWRUN) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java]: Is their a way to call this from Daemon thread.?

File: api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java [hunk 8]
```diff
@@ -49,13 +54,23 @@ public void willStart() {
...
 			    LocationBasedAccessConstants.PERSONATTRIBUTETYPE_UUID);
 	    createGlobalPropertyForLocationAttribute(LocationBasedAccessConstants.REF_APP_LOCATION_USER_PROPERTY_NAME,
 			    LocationBasedAccessConstants.LOCATION_USER_PROPERTY_NAME);
 	    log.info("Location Based Access Control Module started");
     }
 
+	private void generatePrivilegesForLocations() {
+		UserService userService = Context.getUserService();
+		List<Location> openMrsLocations = Context.getLocationService().getAllLocations(false);
+		for(Location location : openMrsLocations) {
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/LocationBasedAccessActivator.java]: fix the code style issue

File: api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java [hunk 9]
```diff
@@ -80,6 +85,31 @@ public static String getUserAccessibleLocationUuid(User authenticatedUser) {
         return accessibleLocationUuid;
     }
 
+    /**
+     * Used to get the accessible locations for the user. It will first get from the user property. if the user
+     * property is not available, then check for the locations having user privilege. If both are not available then return null.
+     * @param authenticatedUser Authenticated user
+     * @return accessible Location uuid
+     */
+    public static List <String> getUserAccessibleLocationUuidsWithUserPrivileges(User authenticatedUser){
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java]: fix the code style issue

File: api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java [hunk 10]
```diff
@@ -80,6 +85,31 @@ public static String getUserAccessibleLocationUuid(User authenticatedUser) {
...
+        List <String> accessibleLocationUuids = new ArrayList<String>();
+        String accessibleLocationUuid = authenticatedUser.getUserProperty(LocationBasedAccessConstants.LOCATION_USER_PROPERTY_NAME);
+        if (StringUtils.isBlank(accessibleLocationUuid)) {
+            List<Location> openMrsLocations = Context.getLocationService().getAllLocations();
+            for(Location location : openMrsLocations) {
+                if(authenticatedUser.hasPrivilege("LocationAccess "+location.getName())){
+                    accessibleLocationUuids.add(location.getUuid());
+                }
+            }
+        }else {
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java]: fix the code style issue

File: api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java [hunk 11]
```diff
@@ -80,6 +85,31 @@ public static String getUserAccessibleLocationUuid(User authenticatedUser) {
...
+    public static List <String> getUserAccessibleLocationUuidsWithUserPrivileges(User authenticatedUser){
+        if (authenticatedUser == null) {
+            return null;
+        }
+        List <String> accessibleLocationUuids = new ArrayList<String>();
+        String accessibleLocationUuid = authenticatedUser.getUserProperty(LocationBasedAccessConstants.LOCATION_USER_PROPERTY_NAME);
+        if (StringUtils.isBlank(accessibleLocationUuid)) {
+            List<Location> openMrsLocations = Context.getLocationService().getAllLocations();
+            for(Location location : openMrsLocations) {
+                if(authenticatedUser.hasPrivilege("LocationAccess "+location.getName())){
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java]: fix the code style issue

File: api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java [hunk 12]
```diff
@@ -80,6 +85,31 @@ public static String getUserAccessibleLocationUuid(User authenticatedUser) {
...
+     */
+    public static List <String> getUserAccessibleLocationUuidsWithUserPrivileges(User authenticatedUser){
+        if (authenticatedUser == null) {
+            return null;
+        }
+        List <String> accessibleLocationUuids = new ArrayList<String>();
+        String accessibleLocationUuid = authenticatedUser.getUserProperty(LocationBasedAccessConstants.LOCATION_USER_PROPERTY_NAME);
+        if (StringUtils.isBlank(accessibleLocationUuid)) {
+            List<Location> openMrsLocations = Context.getLocationService().getAllLocations();
+            for(Location location : openMrsLocations) {
```
Comment (suthagar23) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/utils/LocationUtils.java]: fix the code style issue

File: api/src/test/java/org/openmrs/module/locationbasedaccess/aop/LocationUtilsTest.java [hunk 13]
```diff
@@ -190,4 +198,50 @@ public void getUserAccessibleLocationUuid_shouldReturnEmptyStringIfNoSessionLoca
 		assertEquals("", userAccessibleLocationUuid);
 	}
 
+	@Test
+	public void getUserAccessibleLocationUuidsWithUserPrivileges_shouldReturnNullIfNoUserPropertyAndRoles(){
```
Comment (suthagar23) [inline api/src/test/java/org/openmrs/module/locationbasedaccess/aop/LocationUtilsTest.java]: fix the code style issue

**Pull request merged**